feat(watchArray): new function

w1ndy opened the pull request:
<!-- Thank you for contributing! -->

### Description

This PR adds a new watcher helper ``watchListChanges``. This function monitors the changes made to a list and notifies the callback with the added and removed elements.

### Additional context

<!-- e.g. is there anything you'd like reviewers to focus on? -->

---

### What is the purpose of this pull request? <!-- (put an "X" next to an item) -->

- [ ] Bug fix
- [x] New Feature
- [ ] Documentation update
- [ ] Other

### Before submitting the PR, please make sure you do the following

- [x] Read the [Contributing Guidelines](https://github.com/vueuse/vueuse/blob/main/CONTRIBUTING.md).
- [x] Read the [Pull Request Guidelines](https://github.com/vueuse/vueuse/blob/main/packages/guidelines.md).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [x] Ideally, include relevant tests that fail without this PR but pass with it.

w1ndy commented:
This function is handy when using hooks with v-for:
```html
<div v-for="k in num" :ref="refs.set">{{ k }}</div>
```
```typescript
const refs = useTemplateRefsList<HTMLDivElement>()
watchListChanges(refs, (_new, _old, added) => {
    for (let el of added) {
        useEventListener(el, 'click', () => console.log('clicked!'))
    }
}, { deep: true })
```

antfu commented on packages/shared/watchListChanges/index.md:
```diff
@@ -0,0 +1,28 @@
+category: Watch
+---
+
+# watchListChange
```
Maybe `watchArray`?

antfu commented on packages/shared/watchListChanges/index.test.ts:
```diff
@@ -0,0 +1,64 @@
+    watchListChanges(num, spy, { deep: true })
+    num.value.splice(1, 1, 5, 6, 7)
+    await nextTick()
+    expect(spy).toBeCalledTimes(1)
```
Would be great to also have some test for `reactive([1,2,3])` and `() => arr.value` as watch source.

antfu commented on packages/shared/watchListChanges/index.md:
```diff
@@ -0,0 +1,28 @@
+
+## Usage
+
+Similar to `watch`, but provides the added and removed elements to the callback function. Pass ``{ deep: true }`` if the list is updated in-place with ``push``, ``splice``, etc.
```
```suggestion
Similar to `watch`, but provides the added and removed elements to the callback function. Pass `{ deep: true }` if the list is updated in place with `push`, `splice`, etc.
```

w1ndy commented:
@antfu Thanks for your feedback, I've renamed this function to `watchArray` and added two tests for the reactive and functional watch sources. Could you help review again?

antfu commented on packages/shared/watchArray/index.md:
```diff
@@ -0,0 +1,28 @@
+watchArray(list, (newList, oldList, added, removed) => {
+  console.log(newList) // [1, 2, 3, 4]
+  console.log(oldList) // [1, 2, 3]
+  console.log(added) // [4]
+  console.log(removed) // []
```
Can `added` and `removed` also have meta data like `[{ value: 4, index: 3 }]`?
